main.go

```diff
@@ -0,0 +1,129 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	kustomizeapi "sigs.k8s.io/kustomize/api/types"
+
+	stacksv1alpha1 "github.com/crossplaneio/resourcepacks/api/v1alpha1"
+	"github.com/crossplaneio/resourcepacks/pkg/controllers"
+	"github.com/crossplaneio/resourcepacks/pkg/operations/kustomize"
+)
+
+const (
+	KustomizeEngine = "kustomize"
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+func main() {
+	var (
+		// top level app definition
+		app = kingpin.New(filepath.Base(os.Args[0]), "An open source multicloud control plane.").DefaultEnvars()
+		//debug      = app.Flag("debug", "Run with debug logging.").Short('d').Bool()
+
+		// The default controller mode.
+		controllerCmd = app.Command(filepath.Base(os.Args[0]), "An open source multicloud control plane.").Default()
+
+		// Configuration for the reconciler.
+		reconcileCmd                = controllerCmd.Command("reconcile", "Reconcile a Custom Resource")
+		templateStackNameInput      = controllerCmd.Flag("template-stack-name", "Name of the TemplateStack custom resource.").Required().String()
+		templateStackNamespaceInput = controllerCmd.Flag("template-stack-namespace", "Namespace of the TemplateStack custom resource").String()
+	)
+	cmd := kingpin.MustParse(app.Parse(os.Args[1:]))
+
+	kingpin.FatalIfError(clientgoscheme.AddToScheme(scheme), "could not register client-go scheme")
+	kingpin.FatalIfError(stacksv1alpha1.AddToScheme(scheme), "could not register stacks group scheme")
+
+	switch cmd {
+	case reconcileCmd.FullCommand():
+		mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+			Scheme: scheme,
+			Port:   9443,
+		})
+		kingpin.FatalIfError(err, "unable to start manager")
+
+		ts := &stacksv1alpha1.TemplateStack{
+			ObjectMeta: v1.ObjectMeta{
+				Name:      *templateStackNameInput,
+				Namespace: *templateStackNamespaceInput,
+			},
+		}
+		kingpin.FatalIfError(getTemplateStack(ts), "could not fetch the TemplateStack object")
+
+		var options []controllers.ResourcePackReconcilerOption
+		switch ts.Spec.Behavior.EngineConfiguration.Type {
+		case KustomizeEngine:
+
+			// TODO(muvaf): investigate a better way to convert *Unstructured to *Kustomization.
+			kustomizationYAML, err := yaml.Marshal(ts.Spec.Behavior.EngineConfiguration.Kustomization)
+			kingpin.FatalIfError(err, "cannot marshal kustomization object")
+			kustomization := &kustomizeapi.Kustomization{}
+			kingpin.FatalIfError(yaml.Unmarshal(kustomizationYAML, kustomization), "cannot unmarshal into kustomization object")
+
+			options = append(options, controllers.WithTemplatingEngine(kustomize.NewKustomizeEngine(kustomization,
+				kustomize.WithResourcePath(ts.Spec.Behavior.Source.Path),
+				kustomize.AdditionalOverlayGenerator(kustomize.NewPatchOverlayGenerator(ts.Spec.Behavior.EngineConfiguration.Overlays)),
+			)))
+		}
+		gvk := schema.FromAPIVersionAndKind(ts.Spec.Behavior.CRD.APIVersion, ts.Spec.Behavior.CRD.Kind)
+		controller := controllers.NewResourcePackReconciler(mgr, gvk, options...)
```

Comment:
I feel a little weird about all this happening in the `main.go` instead of in a different file and method that's decoupled from the `main.go`, but I'm guessing it's happening here because this is an early pass and we're just trying to get something working at this point

Reply:
Yes, possibly another file could work but the calls won't be simpler and they will happen in `main.go` since we have to do all this in before starting the controller.

One thing I've been considering is to set up the reconciler in each reconcile, meaning we will only start the controller but in each reconciler we'll set engine type, patchers etc. But that will complicate the logic to some extent. The downside of the current approach is that when you change something in `StackDefinition`, controller will need to be restarted. But honestly, I don't think `StackDefinition` is something ever changes.